### PR TITLE
datapath: preserve tail call maps in between replacing interface program pairs

### LINKED ISSUE #16561: bpf: cilium-map-migrate might unpin in-use tailcall map
After adding a new taill call to e.g. `CILIUM_MAP_CALL` (as it's been done in https://github.com/cilium/cilium/pull/14138/files) and then upgrading Cilium, there will be a small window when packets will be dropped due to `DROP_MISSED_TAIL_CALL`.

When we load e.g. bpf_host on a native device (e.g. eth0), the following is executed by the cilium-agent:
```
$ cilium-map-migrate -s 42/bpf_netdev_eth0.o
$ tc filter replace dev eth0 ingress prio 1 handle 1 bpf da obj 42/bpf_netdev_eth0.o sec from-netdev
$ cilium-map-migrate -e 42/bpf_netdev_eth0.o -r 0

$ cilium-map-migrate -s 42/bpf_netdev_eth0.o
# renames /sys/fs/bpf/tc/globals/cilium_calls_netdev_00666{,:pending} <-- it's being used by bpf_netdev_eth0[from-netdev]@eth0[ingress]
$ tc filter replace dev eth0 egress prio 1 handle 1 bpf da obj 42/bpf_netdev_eth0.o sec to-netdev
# creates /sys/fs/bpf/tc/globals/cilium_calls_netdev_00666 <-- it's being used by bpf_netdev_eth0[to-netdev]@eth0[egress]
$ cilium-map-migrate -e 42/bpf_netdev_eth0.o -r 0
# deletes /sys/fs/bpf/tc/globals/cilium_calls_netdev_00666:pending (used by bpf_netdev_eth0[from-netdev])
```

Because BPF maps of the `BPF_MAP_TYPE_PROG_ARRAY` type are removed immediately once they are unpinned (regardless of the refcount), we get the missed tail calls on the ingress of eth0.

This became obvious after analyzing test failures in the PR https://github.com/cilium/cilium/pull/13032#issuecomment-861531511.

### PR DESCRIPTION
**Pull request by ti-mo**
To prevent unpinning tail call maps after replacing only one half of an interface's programs, e.g. `from-netdev` and `to-netdev`, allow the caller to defer finalizer operations until a batch of interfaces have been replaced.

This avoids short windows of packet drops due to missed tail calls.

Fixes #16561

Also fixed a bug where we specified the wrong bpffs path to the migration code. This causes migrations to not be executed in the currently-released version of 1.11, meaning resizing maps causes endpoints regeneration failures.

```release-note
Preserve tail call maps during resize to prevent drops during agent upgrade
```

### CODE REVIEW

**Review by kkourt — APPROVED** [review]
LGTM.

I'm not 100% sure about  not garbage collecting  the files, but if this only happens when the map size is changed (which, I guess, does not happen that often) I cannot think of a strong argument against it. 

That being said, could we do it an easy version of GC where we keep only _one_ backup directory? That is, after the backup is done we look for older backups for the same map and delete them.

**Review by pchaigno — APPROVED** [review]
LGTM. Thanks!

Did you already test this by adding random tail call somewhere and executing K8sUpdates?

> I'm not 100% sure about not garbage collecting the files, but if this only happens when the map size is changed (which, I guess, does not happen that often) I cannot think of a strong argument against it.

If it becomes an issue, we could maybe rely on the file creation timestamps in bpffs to determine which backups have been made a very long time ago and are safe to delete. But I really doubt Kubernetes nodes have a lifespan long enough for this to become an issue.

> That being said, could we do it an easy version of GC where we keep only one backup directory? That is, after the backup is done we look for older backups for the same map and delete them.

I don't think we can even assume datapath logic n-1 is fully loaded when we start loading datapath logic n. So there could still be bpf_lxc programs using tail call maps from the n-2 version... or even before :-/

**Comment by ti-mo** [comment]
> Did you already test this by adding random tail call somewhere and executing K8sUpdates?

I've tested this by creating a prog array of a different size manually and running the migration commands. Not sure how to work with `K8sUpdates`.

> But I really doubt Kubernetes nodes have a lifespan long enough for this to become an issue.

We can't really assume all k8s nodes are short-lived, as baremetal deployments are a thing as well. However, we only need to resize these maps once in a blue moon and they're rather small. If we do resize them, we should take care to grow them sufficiently (e.g. x1.5) so that adding future tail calls does not require a map migration.

**Comment by ti-mo** [comment]
Marked as ready, only known unrelated flakes are hit. This should suffice for 1.11, followups can happen in case any issues become apparent with this approach.

File: pkg/bpf/bpffs_migrate.go
Comment (joestringer) [inline pkg/bpf/bpffs_migrate.go]: When this says "Never", what exactly does that mean? Does that mean we will leak these maps upon each upgrade? How will they get cleaned up?

File: pkg/bpf/bpffs_migrate.go
Comment (joestringer) [inline pkg/bpf/bpffs_migrate.go]: Do we really need to log this at info level? Seems like a very specific implementation detail which means nothing to users.

File: pkg/bpf/bpffs_migrate.go
Comment (joestringer) [inline pkg/bpf/bpffs_migrate.go]:
```suggestion
		return fmt.Errorf("unable to create bpffs backup directory: %w", err)
```

File: pkg/bpf/bpffs_migrate.go
Comment (ti-mo) [reply pkg/bpf/bpffs_migrate.go]: :wave: Rationale here: https://github.com/cilium/cilium/issues/16561#issuecomment-938534720

File: pkg/bpf/bpffs_migrate.go
Comment (ti-mo) [reply pkg/bpf/bpffs_migrate.go]: This feels like a crucial event to have in the logs as it's only performed during upgrades when map properties have changed. Especially given the troubleshooting history that led to this PR, being able to observe this event (or a lack thereof) could be important in some situations.

I'll leave this up to @brb or @borkmann to decide, but will leave it in for now.

**Comment by jrajahalme** [comment]
Is this superseding #13032?

**Comment by ti-mo** [comment]
> Is this superseding #13032?

Yep, I think it does!

File: pkg/bpf/bpffs_migrate.go
Comment (joestringer) [reply pkg/bpf/bpffs_migrate.go]: I couldn't quite tell from the linked comment, I guess admittedly I probably would need to spend a bit more time understanding. But if the answer is really "never", then I think we need to fix this. Otherwise we are forever consuming resources in the kernel based on the number of times we restart Cilium.

Comment (ti-mo) [reply pkg/bpf/bpffs_migrate.go]: > Otherwise we are forever consuming resources in the kernel based on the number of times we restart Cilium.

As mentioned in the linked comment: `Prog array migration only happens on Cilium up/downgrade, which is not supposed to happen all too frequently`. For example, https://github.com/cilium/cilium/pull/17758 proposes to bump `CILIUM_CALL_SIZE` from 25 to 27, which will result in that tail call map being recreated with the new size, once, when the agent is upgraded. This leaves one lingering prog array with 25 max entries on the system.

The alternative is periodically scanning all progs in the kernel (which can run into the thousands on busy nodes) for references to this old map. This process might need to go on forever until all pods using the map have been killed/rotated. Scanning takes (at least) one syscall per prog in the kernel.

Not accounting for repeated up/downgrades that will create one backup map per tick, I'm not convinced that adding a GC is worth the development and runtime execution overhead given the small size of these maps. Alternatively, we could always document that these map pins can be safely deleted if all managed pods have been rotated on a node post-upgrade.

Comment (joestringer) [reply pkg/bpf/bpffs_migrate.go]: Is there a solution where we just delete this `backup` directory, thereby unpinning after some time? Or track `backup1` and `backup2` and then upon upgrade, we always move the current ones to `backup1`, move the old `backup1` to `backup2` and make sure the old `backup2` are deleted?

Comment (ti-mo) [reply pkg/bpf/bpffs_migrate.go]: > Is there a solution where we just delete this `backup` directory, thereby unpinning after some time?

Note that deleting an in-use tail call map results in dropped packets. Establishing the `after some time` heuristic is the problem here. For example, interrupted connectivity 10 mins after an upgrade would be even more surprising than interruptions during an upgrade.

> Or track backup1 and backup2 and then upon upgrade, we always move the current ones to backup1, move the old backup1 to backup2 and make sure the old backup2 are deleted?

Not sure how we would track whether `backup1` is current, old or new without keeping persistent state on the host. It's not clear what exactly you're proposing here. As Paul mentioned above, it's not even guaranteed that all pods created using Cilium version n-2 have died before the system was upgraded to n-1. There is always a risk of interruptions when blindly (without refscanning) unpinning a tail call map.

Comment (joestringer) [reply pkg/bpf/bpffs_migrate.go]: :+1: indeed I agree that "after some fixed time" is not a reasonable solution.

One thing I don't understand is why we don't know if the map is still in use or not. In general, upon startup Cilium is intended to replace all running eBPF programs (previously installed by earlier Cilium) with new programs to ensure that the datapath is now running the new datapath implementation code. We should know exactly all of the possible points that should be configured and attached. Broadly this consists of the core programs eg attached to netdevices, sockets etc. plus per-endpoint programs. Then if there is anything else left over that we didn't migrate to the new version, we should be able to safely remove it. We already persist the information necessary to determine this.

Comment (pchaigno) [reply pkg/bpf/bpffs_migrate.go]: > One thing I don't understand is why we don't know if the map is still in use or not.

We can know that, but we'd have to parse the xlated BPF programs from the kernel to search for all references to the tail call map.

> +1 indeed I agree that "after some fixed time" is not a reasonable solution.

It is not because you're both talking minutes. Are your concerns with that approach still valid if we delete maps older than a week?

Comment (joestringer) [reply pkg/bpf/bpffs_migrate.go]: > We can know that, but we'd have to parse the xlated BPF programs from the kernel to search for all references to the tail call map.

Where I'm coming from is, we have a set of known intended state in the agent, and we will then conform all the hook points that we control to the intended state. This approach says "I don't want to analyse the current situation from the kernel perspective and handle all possible references that exist", it says "If there are known paths on the filesystem for pinned programs, we will move those out of the way and then instantiate the new programs in the old places. When I am done, everything should be in the correct state and there should be no remaining references to the old programs.". If we are still missing cases there, then we aren't doing enough up-front tracking of the intended state of the system.

> It is not because you're both talking minutes. Are your concerns with that approach still valid if we delete maps older than a week?

I think ultimately this is worse. Rather than "hmm, connectivity for my pod breaks after N minutes following upgrade", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not. I'd much rather do the wrong thing early with some possibility of understanding the trigger rather than doing the wrong thing much later and have no idea why it occurred. (Of course, I'd rather do the right thing but I'm covering worst-case scenarios here :sweat_smile: )

Comment (pchaigno) [reply pkg/bpf/bpffs_migrate.go]: > Rather than "hmm, connectivity for my pod breaks after N minutes following upgrade", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not.

Uh, wouldn't this only happen if the endpoints fail to be regenerated for over a week? I think we have more serious issues if that happens :grimacing:

Comment (joestringer) [reply pkg/bpf/bpffs_migrate.go]: Is that true? That feels like it frames the resource cleanup question much more nicely.

If everything succeeds regeneration, delete the `backup` directory. Otherwise we need to keep it around to keep the endpoints successfully connected, and hopefully the warning signs from regeneration will reach the user and they'll roll back to a known good state, ie previous version. In that particular case there would still be a `backup` directory on the system until they next upgrade, but at that point we could just remove/overwrite those `backup` pins and hopefully have better luck the second time (presumably with a new release of the target upgrade version).

File: pkg/bpf/bpffs_migrate.go
Comment (brb) [reply pkg/bpf/bpffs_migrate.go]: I'd keep the log statement at least until we have enough confidence that the new map migrate infrastructure is sound.

**Comment by ti-mo** [comment]
As a quick follow-up, we've agreed a while back to implement a GC that would:
1. scan all prog arrays in `backup/`
2. iterate through all BPF programs on the system
3. scan all existing BPF programs for references to any prog arrays
4. if no references are found to a given prog array, unpin it

Starting work on this this week.

**Comment by ti-mo** [comment]
After some offline discussion, we'll move forward with removing `backup/` (where 'old' prog arrays are moved to) when a certain set of programs have been successfully regenerated.

- Will unblock existing PRs
- Should make upgrade tests pass on changes that grow prog arrays
- Should be an improvement in the majority of cases, whereas now there is full connectivity interruption
- Can be further iterated on with changes to the loader pipeline throughout 1.12/1.13
- Should allow test suite of https://github.com/cilium/cilium/pull/17758 to pass
- Can be reasonably backported to 1.11 to improve up/downgrade behaviour

Known issues with this approach:

- Pods scheduled during upgrades might still be given an old map that gets cleared soon after the Pod has started. The general advice is to halt pod scheduling during upgrades, but we can document this and treat this as a known issue.
- What if the agent is still busy applying datapath version n-1 that happens to include another resize?
  -> This should be rare and might be something we can ignore, eventually everything will be regenerated and reconnected.

This should improve the total loss of connectivity during upgrades within a reasonable timeframe.

**Comment by ti-mo** [comment]
I couldn't find a suitable point to reliably remove a bpffs `backup/` directory, but the main impact seems to be when unpinning a map in between replacing both ends of a program pair of an interface, e.g. replace eth0:from-netdev, <unpin>, replace eth0:to-netdev.

The ability to defer the unpinning operation until both ends have been replaced has been added to `replaceDatapath` in the form of a finalizer. It can either be called immediately, or, when called in a loop, deferred.

I also spotted a bug we unfortunately shipped in 1.11, the wrong bpffs path is specified to the migration procedure. The migrations run through init.sh are unaffected. It's not clear what the impact of this is, but I would expect things to break loudly, as the verifier would likely complain about incompatible maps.

File: pkg/datapath/loader/netlink.go
Comment (pchaigno) [inline pkg/datapath/loader/netlink.go:85]: Nice catch!

Maybe one explanation why we didn't get more bug reports from this error is because we may not have changed that many maps between v1.10 and v1.11. And if we did, it may have been only for advanced features, disabled by default (so not covered in upgrade test).

Comment (pchaigno) [inline pkg/datapath/loader/netlink.go:85]: Are we confident enough about the second fix to backport it to v1.11 straight away? That would mean it will be included in v1.11.2, which larger production users are likely to upgrade to. If not, then I think we'll need to move the first bugfix to a separate PR, to be able to backport it now without backporting the more impactful change from the second commit.

cc @joestringer for a maintainer's perspective.

File: pkg/datapath/loader/loader.go
Comment (pchaigno) [inline pkg/datapath/loader/loader.go]: I thought the goal of this callback function was to only remove maps once all BPF programs for the interface have been replaced. But if `ep.RequireEgressProg()` is true below, then we haven't yet replaced the egress program for the interface when we execute this `finalize`.

**Review by pchaigno — COMMENTED** [review]
Nice catch on both bugs! In particular, the second fix is definitely simpler than other solutions we had in mind. Not sure how we missed that solution before :facepalm: 

(If we run into any issue later on with the second bugfix, I think another approach could be to simply separate the per-endpoint tail call map into ingress and egress tail call maps. AFAIK, it doesn't actually need to be shared between both directions.)

My one concern with this is on how we want to backport, as outlined below. Are we confident enough in the second bugfix to backport it right away? Did you perform any upgrade tests other than the CI's? How did you check that it is fixing #16561?

File: pkg/datapath/loader/netlink.go
Comment (ti-mo) [reply pkg/datapath/loader/netlink.go:85]: I'm open to carving out the first fix (bpffs path) into a separate PR. I think it's rather critical given it can deadlock datapath regeneration (or even prevent agent start?) when maps get changed.

The second (deferred map removal) is mostly oriented towards 1.12, but I was considering the case where we need to add and backport a tail call as an 'emergency' if a complexity issue arises that we cannot solve any other way. It's completely up to you to; perhaps we should only backport this fix if/when such a situation actually arises.

**Comment by ti-mo** [comment]
> Did you perform any upgrade tests other than the CI's? How did you check that it is fixing #16561?

I did not perform my own connectivity tests (I was waiting for some feedback/reviews first), but I did observe the deferred map removal behaviour by watching changes to the bpffs dir during an upgrade, which seems to behave as expected. I'll do some connectivity tests and report back.

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [reply pkg/datapath/loader/loader.go]: Good catch. I'll defer both finalizers here.

**Comment by ti-mo** [comment]
I've just done some local upgrade tests, both with and without the second (defer) patch.

- Minikube running 4.19.202 under KVM
- Deploying an agent Pod flipping `CILIUM_CALL_SIZE` between 25 and 26 to force prog array re-creation
- Running `cilium monitor --type=drop` in the agent container as soon as the new pod starts
- Running `hey` from my host (not within the VM) against a single nginx Pod inside the cluster at around 10k req/s

There is no noticeable loss of connectivity during the upgrades not even without the patch, and no monitor logs of missed tail calls.

Latency distribution looks good, too. p99 is 0.0034s across 1M HTTP calls.

```
  0.077 [999857]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.155 [37]	|
  0.232 [45]	|
... some buckets omitted ...
  0.774 [3]	|
```

@brb How can I reproduce this best?

File: pkg/datapath/loader/netlink.go
Comment (ti-mo) [reply pkg/datapath/loader/netlink.go:85]: On second thought, I think this is good to go. I don't see this as a particularly invasive or risky change, it just defers cleanup to be slightly later.

File: pkg/datapath/loader/netlink.go
Comment (joestringer) [inline pkg/datapath/loader/netlink.go]: The comment on line 99 states that if the load fails, the original maps will be repinned to their original paths. However, in that case (lines 104-106) the code only sets `revert = true` and then returns a `nil` finalizer. As a result, `bpf.FinalizeBPFFSMigration()` does not get called.

In the successful case, I think that we need to defer the map migration finalization until after all of the programs have been loaded, so that case makes sense to pass the finalizer back to the caller.

In the error case where the program cannot be loaded, I think it is OK to just run the contents of this finalizer function directly from here with `revert=true` and pass back a nil finalizer (like the code on line 106 does already).

File: pkg/datapath/loader/netlink.go
Comment (joestringer) [reply pkg/datapath/loader/netlink.go:85]: >  I think it's rather critical given it can deadlock datapath regeneration (or even prevent agent start?) when maps get changed.

Can you describe this in a bit more detail, also include it in the commit message? It's not immediately obvious to me which conditions would cause this kind of situation or even that the first patch is resolving such a serious bug.

> On second thought, I think this is good to go. I don't see this as a particularly invasive or risky change, it just defers cleanup to be slightly later.

:heavy_plus_sign: Assuming we resolve the other feedback, this should be fairly low risk.

File: pkg/datapath/loader/loader.go
Comment (joestringer) [inline pkg/datapath/loader/loader.go:326]: If I follow right, the approach here is to ensure that both the ingress+egress programs for a given endpoint are completely reloaded before unpinning the old references. That makes some sense to me.

Similarly for the `reloadHostDatapath` case above, the programs for all interfaces are loaded first before unpinning the old map references. That also makes sense. :+1: 

I recall that some of the discussion previously was around what happens when we reload base programs, and how that could impact upgrade. Looking through, it seems like we've managed to sidestep that question for now while resolving the immediate concerns around upgrade. Is that right? The approach here seems like a good incremental improvement and allows us to defer dealing with the `init.sh` conversion and this other potential datapath disruption question until the go migration work matures.

(no changes necessary, this is more just me checking if I am following correctly)

File: pkg/bpf/bpffs_migrate.go
Comment (joestringer) [inline pkg/bpf/bpffs_migrate.go]: nit, Maybe a dumb question here, but the commit message says:

> The `tc/globals/` part of the path was missing, and the bpffs root was being scanned instead. Prevent this error from going unnoticed in the future by explicitly checking if the map directory exists.

If the directory path being passed in was `/sys/fs/bpf` instead of `/sys/fs/bpf/tc/globals`, then won't the directory check pass regardless of the bug? Also if we're part way through migrating these maps, then presumably the directories already exist somewhere, so why is it helpful to make these error checks here?

If there's a reasonable reason, then I'm fine with keeping these, my above questions are really just assessing what kind of difference they make and whether we need these at all (here and in `finalizeMap()`).

**Review by joestringer — CHANGES_REQUESTED** [review]

File: pkg/datapath/loader/netlink.go
Comment (ti-mo) [reply pkg/datapath/loader/netlink.go]: Oops, good catch. I opted to get rid of the bool and call finalize directly when handling the error. I also specified this behaviour in the `replaceDatapath` docstring.

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [reply pkg/datapath/loader/loader.go:326]: Yup, I think we're on the same page here. Could you specify what you mean by 'base programs'?

File: pkg/bpf/bpffs_migrate.go
Comment (ti-mo) [reply pkg/bpf/bpffs_migrate.go]: I see this more as a courtesy to the user/caller, e.g. when `/sys/fs/bpf` is not mounted or `tc/globals` not present for some reason.

Actually, now I think of it, this might do more harm than good if migrations are run before iproute2 had the chance to create `tc/globals`. Missing pins are ignored (system already in desired state), but missing `tc/globals` is an error. I'll remove the checks, they don't really belong there.

File: pkg/datapath/loader/loader.go
Comment (joestringer) [reply pkg/datapath/loader/loader.go:326]: Sure thing. Base programs are the "global" programs that apply once for the node, like bpf_host, bpf_overlay, etc. Structurally in the code, when Cilium wants to update these programs, it calls into the datapath [Reinitialize()](https://github.com/cilium/cilium/blob/477cccaa2afad1aaa49e29dec80952fc050844c7/pkg/datapath/loader/base.go#L237) function. It's typically pretty rare to reinitialize the datapath like this during normal Cilium operations.

This is in contrast to the per-endpoint `bpf_lxc` programs which are cloned per-endpoint and are dynamically configured based on creation of workloads on the nodes. These endpoint BPF programs are installed via the endpoint-specific code in `pkg/endpoint/bpf.go`. Thes actual program reload for endpoints goes via either:
* `CompileAndLoad()`: unusual; typically only triggered by `cilium endpoint regenerate` CLI,
* [CompileOrLoad()](https://github.com/cilium/cilium/blob/477cccaa2afad1aaa49e29dec80952fc050844c7/pkg/endpoint/bpf.go#L699): typical case on startup, or
* `ReloadDatapath()`: used when updating maps for individual endpoints, does not trigger recompilation of ELFs and does not reload the global programs.

**Comment by ti-mo** [comment]
@joestringer Addressed your feedback, please take another look.

I've specified the following in the first commit:

> This would lead to endpoints continuously failing to regenerate at runtime when map definitions in the ELF differ from their pinned counterparts, e.g. during an agent upgrade.
> 
> For example, when resizing CALLS_MAP and trying to replace a program, the old map will be loaded from bpffs, while the program expects to see a map of a different capacity:
> 
> ```
> ~ tc filter replace dev cilium-probe ingress bpf da obj bpf/bpf_lxc.o sec from-container
> libbpf: couldn't reuse pinned map at '/sys/fs/bpf/tc//globals/test_cilium_calls_65535': parameter mismatch
> libbpf: map 'test_cilium_calls_65535': error reusing pinned map
> libbpf: map 'test_cilium_calls_65535': failed to create: Invalid argument(-22)
> libbpf: failed to load object 'bpf/bpf_lxc.o'
> Unable to load program
> ```

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [reply pkg/datapath/loader/loader.go:326]: @joestringer Thanks, I thought these could only regenerate on agent startup, TIL `endpoint regenerate`.

File: pkg/datapath/loader/netlink.go
Comment (joestringer) [inline pkg/datapath/loader/netlink.go:114]: FYI there's also `log.WithFields(logrus.Fields{ ... })` to specify multiple fields at once with one call. This is a more common style elsewhere in the codebase, but also this doesn't make any meaningful difference so don't worry about changing this one.

**Review by joestringer — APPROVED** [review]
LGTM, thanks again for the fix.

**Comment by pchaigno** [comment]
All tests are passing and reviews are in. Marking ready to merge.

**Pull request merged**
